invert linetable representation

vtjnash opened the pull request:
Previously, our linetables (similar to LLVM) represented line information as a linked list from callee via inlined_at up to the original information. This requires many copies of this information to be created. Instead we can take advantage of the necessary existence of the line table from the child to flip this chain of information and instead make each statement be a table describing (for each IR instruction):
`(current line number, (index into edges, index into edges statements))` plus a table of all edges, plus a table with the original line numbers from the parser, plus the file name. This is all packed into the struct

    struct DebugInfo
        def::Union{Method,MethodInstance,Symbol}
        linetable::Union{Nothing,DebugInfo}
        edges::SimpleVector{DebugInfo}
        codelocs::String
    end

Which is described in doc/src/devdocs/ast.md for what each field means and look at stacktraces.jl or compiler/ssair/show.jl to look at how to decode and interpret this information.

For the sysimage, this saves several megabytes (about 113 MB -> 110 MB) and about 5% of the stdlib pkgimages (294 MB -> 279 MB).

It also now happens to have the full type information for the inlined functions. Now if you create an `IRShow.DILineInfoPrinter` with `showtypes=true`, it can print that information when printing IR.

Opening as draft, although it is hopefully ready for review, since it appears there will be an update needed to keep Revise working with the new format for this info.

```
julia> @eval Base.IRShow DILineInfoPrinter(debuginfo, def) = DILineInfoPrinter(debuginfo, def, true)
DILineInfoPrinter (generic function with 2 methods)

julia> (@code_typed 1 + 1.0)[1]
CodeInfo(
    @ promotion.jl:425 within `+`
   ┌ invoke MethodInstance for promote(::Int64, ::Float64)
   │ @ promotion.jl:396 within `promote`
   │┌ invoke MethodInstance for Base._promote(::Int64, ::Float64)
   ││ @ promotion.jl:373 within `_promote`
   ││┌ invoke MethodInstance for convert(::Type{Float64}, ::Int64)
   │││ @ number.jl:7 within `convert`
   │││┌ invoke MethodInstance for Float64(::Int64)
   ││││ @ float.jl:221 within `Float64`
1 ─││││ %1 = Base.sitofp(Float64, x)::Float64
│  └└└└
│  ┌ invoke MethodInstance for +(::Float64, ::Float64)
│  │ @ float.jl:460 within `+`
│  │ %2 = Base.add_float(%1, y)::Float64
│  └
└──      return %2
)
```

vchuravy commented on base/compiler/ssair/ir.jl:
```diff
@@ -1570,6 +1630,8 @@ function resize!(compact::IncrementalCompact, nnewnodes::Int)
     return compact
 end
 
+const NoLineUpdate = (Int32(0), Int32(0), Int32(0))
```
Could we use a `struct` here? It's not clear what the three entries are for.

JeffBezanson commented on src/jltypes.c:
```diff
@@ -3051,10 +3051,10 @@ void jl_init_types(void) JL_GC_DISABLED
-                        jl_perm_symsvec(5, "module", "method", "file", "line", "inlined_at"),
-                        jl_svec(5, jl_module_type, jl_any_type, jl_symbol_type, jl_int32_type, jl_int32_type),
-                        jl_emptysvec, 0, 0, 5);
+        jl_new_datatype(jl_symbol("LegacyLineInfoNode"), core, jl_any_type, jl_emptysvec,
```
I don't think we should rename this, especially if we are still using it. Maybe it can be merged with `LineNumberNode` though?

vtjnash replied:
We aren't really still using it though. It appears briefly in the output from lowering until we can wrap it in the output (also from lowering) of a CodeInfo, at which point we destroy this and allocate the forward edges structure--which becomes the only types the user should see now. It will probably make more sense later to have lowering emit the new format directly--at which point this placeholder type can be deleted the rest of the way

JeffBezanson replied:
Deleting it will be fine, but as long as it exists we should just leave it alone.

vtjnash replied:
It looks like we still needed this exact type in Core to support serialization of legacy code, so I also now define the user-accessible type in boot.jl directly, with the hope that we remove this object soon (as soon as we can teach lowering not to use it as an intermediate representation)

JeffBezanson commented:
I like this; feels like a real improvement. Just need to figure out some details and how much we care about compatibility of code that handles IR.

vtjnash commented:
> compatibility of code that handles IR.

Thanks! As long as code has been using the Compiler APIs, they should not be much affected. It is rare for code to need to "invent" a novel line number. Inlining does it in some specific ways, but generally code should not do it otherwise, as lines should come from the user, which is what I wanted to better reflect here, compared to what I felt that inlined_at had done previously.

JeffBezanson commented on base/compiler/inferencestate.jl:
```diff
@@ -456,21 +456,16 @@ function should_insert_coverage(mod::Module, src::CodeInfo)
+should_insert_coverage(mod::Module) = false
+function should_insert_coverage(info::DebugInfo)
+    linetable = info.linetable
+    linetable === nothing || (should_insert_coverage(should_insert_coverage) && return true)
```
```suggestion
    linetable === nothing || (should_insert_coverage(linetable) && return true)
```
?

timholy commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+    the current function for metadata. The codelocs line number also becomes an index into
+    this codelocs instead of being a line number itself, as described below.
+
+  * `edges` : Vector of the unique DebugInfo for every inlined function
```
Which direction? Is it the list of things that this inlines into, or the list of things that are inlined into this?

vtjnash replied:
Inlined into this. These are simple forward edges now (things that are called), so this structure walk is now the same as a call-graph walk.

timholy commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+
+  * `edges` : Vector of the unique DebugInfo for every inlined function
+
+  * `firstline` (when uncompressed to DebugInfoStream)
```
The rest of this material is excellent, presuming that you clarify `edges`

timholy commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+     2. the integer index into edges, giving the DebugInfo inlined there (or zero if there
+        are no edges).
+     3. (if entry 2 is non-zero) the integer index into edges[].codelocs, giving the
+        recursion point, or zero indicating to use `edges[].firstline` as the line number.
```
what's a recursion point?

vtjnash replied:
Interpreting the inlined information requires repeating these interpretation steps 1, 2, and 3 with this new index as the new program counter in the new debuginfo table reached via this edge.

timholy commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+        recursion point, or zero indicating to use `edges[].firstline` as the line number.
+
+   Special codes include:
+     - (zero, zero, *) : no change to the line number or edges
```
what does "no change" mean? no change from what? Where is the authoritative source then?

vtjnash replied:
It means "no change" from the last line number, in any context, and meaning that a stepping debugger should not stop on this statement. It is the same as a `0` means now in the current linetable

timholy reviewed (commented):
I can't pretend to have read the whole thing carefully, but the overall concept seems great. Merely having the type info available for inlining is a huge win on its own. I'll certainly make any needed changes to the Revise stack when this gets merged.

My comments are mostly about the clarity of the documentation.

vtjnash commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
     See the documentation of `Base.@assume_effects` for more details.
+
+
+How to interpret line numbers in a `CodeInfo` object:
```
Good questions. I will try to answer them inline, and eventually hopefully update the document to clarify these points also (maybe not immediate as I may be on vacation for the next 2 weeks)

Typically these are attached to a `CodeInfo` object in the `debuginfo` field, but they are expected to be referenced from many other places (one of the main points of this PR is that this will be referenced from many places instead of doing copies as much)

vtjnash commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+    edges::SimpleVector{DebugInfo}
+    codelocs::String # compressed data
+end
+mutable struct DebugInfoStream
```
It also might just need a different name (and a note that this is `Core.Compiler.DebugInfoStream`, not a built-in)? It is a mutable version of the previous struct

vtjnash commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+
+  * `def` : where this DebugInfo was defined (the Method, MethodInstance, or file scope, for example)
+
+  * `linetable`
```
That seems a decent name for it, sure

vtjnash commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+
+  * `linetable`
+
+    Another debuginfo that this was derived from. If `def` is not a Symbol, then it replaces
```
"derived" is roughly the opposite of inlining. Being derived from means it started as the same source and then was optimized to form the new object. It exists because otherwise this structure could not quite represent macros (which need to map multiple sources lines to each statement)

vtjnash commented on doc/src/devdocs/ast.md:
```diff
@@ -763,3 +763,61 @@ Boolean properties:
+  * `linetable`
+
+    Another debuginfo that this was derived from. If `def` is not a Symbol, then it replaces
+    the current function for metadata. The codelocs line number also becomes an index into
```
Yes, it allows a function (generally `@generated`) to spoof being an entirely different function (for debuginfo purposes). It is a bit of a process, and is implemented a few times in different ways (chief among them Base.IRShow.buildLineInfoNode and Stacktraces.lookup)

vtjnash commented:
I was testing Revise on this branch earlier to see if I needed to make a PR there, and I think this PR already resolved a lot of the issues with it: https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/606. Looks like there are a few other failing test now (since it type-asserts the kind of something and doesn't know about the new `Base.IRShow.LineInfoNode`). They look minor, but let me know if something isn't clear.

timholy commented:
When we get a better way of tracking "provenance" (e.g., #31162) through lowering, will this make it easier to extend it all the way through the IRCode representation? It seems plausible to me that it will. Someday we should write an IRCode interpreter, I expect that to be much faster than any interpreter we have now. But it's kinda useless if you can't track whatever you're evaluating back to the original source code.

vtjnash commented:
This already does that in fact. Every statement refers back to the IR statement it was originally from, including the inlining info, instead of the line numbers (only the original statement then has the line number). It isn't one-to-one precise though (as various transforms may combine or delete statements), and doesn't alter lowering to track it back to the AST.

timholy commented:
Right. All that sounds great. Looking forward to this and to getting similar things done for lowering too---it will expand the tooling options considerably.

topolarity commented on base/compiler/ssair/ir.jl:
```diff
@@ -261,11 +308,20 @@ Instruction(is::InstructionStream) = Instruction(is, add_new_idx!(is))
+    fldarray = getfield(getfield(node, :data), fld)
+    fldidx = getfield(node, :idx)
+    if fld === :line
+        (fldarray[3fldidx-2], fldarray[3fldidx-1], fldarray[3fldidx-0]) = val::NTuple{3,Int32}
```
Is it worth giving these `(line, edge, edge_line)` triplets a name?

I had to do a fair bit of digging to figure out how to interpret them

vtjnash replied:
Isn't that covered in the documentation here already?

vtjnash committed:
inverted linetable representation

Previously, our linetables (similar to LLVM) represented line
information as a linked list from callee via inlined_at up to the
original information. This requires many copies of this information to
be created. Instead we can take advantage of the necessary existence of
the line table from the child to flip this chain of information and
instead make each statement be a table describing (for each IR
instruction):
`(current line number, (index into edges, index into edges statements))`
plus a table of all edges, plus a table with the original line numbers
from the parser, plus the file name. This is all packed into the struct

    struct DebugInfo
        def::Union{Method,MethodInstance,Symbol}
        linetable::Union{Nothing,DebugInfo}
        edges::SimpleVector{DebugInfo}
        codelocs::String
    end

Which is described in doc/src/devdocs/ast.md for what each field means
and look at stacktraces.jl or compiler/ssair/show.jl to look at how to
decode and interpret this information.

For the sysimage, this saves several megabytes (about 113 MB -> 110 MB)
and about 5% of the stdlib pkgimages (294 MB -> 279 MB).

It also now happens to have the full type information for the inlined
functions. Now if you create an `IRShow.DILineInfoPrinter` with
`showtypes=true`, it can print that information when printing IR.

vtjnash commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/52415#issuecomment-1979733891) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0f88025_vs_fb71a5d/report.md).
